### PR TITLE
Add recipes for Chemfiles

### PR DESCRIPTION
**Pull request by Luthaf**
A C++ library for reading/writing theoretical chemistry trajectories, with Python bindings. See http://chemfiles.github.io/ for more informations. 

I added three recipes: C++ library, Python binding and meta-package. If you prefer having everything in one recipe, I can to it.

### CODE REVIEW

File: recipes/chemfiles-lib/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    - cmake --build . --config release --target install
+    - cmake -DBUILD_SHARED_LIBS=ON .
+    - cmake --build . --config release --target install
+source:
+  fn: chemfiles-{{ version }}.tar.gz
+  url: https://github.com/chemfiles/chemfiles/archive/{{ version }}.tar.gz
+  sha256: {{ sha256 }}
+
+requirements:
+  build:
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: Please add `toolchain`. This configures the compiler to use and how to use it.

File: recipes/chemfiles-lib/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    - cmake
+    - libnetcdf
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: Please pin to `4.4.*` based on our [global pinnings]( https://github.com/conda-forge/conda-forge.github.io/blob/e720eaa23843f76bd7dde7e4af46671eead72447/scripts/pin_the_slow_way.py#L41 ).

File: recipes/chemfiles-lib/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  run:
+    - libnetcdf
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: Please also pin to `4.4.*`.

File: recipes/chemfiles-python/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  commands:
+    - cd $SRC_DIR/tests
+    - python -m unittest discover
+
+requirements:
+  build:
+    - python
+    - enum34  # [py < 34]
+    - numpy
+    - chemfiles-lib ==0.6.0
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Are these all required at build time? Guessing not. Please drop everything other than `python`. Also please add `setuptools` as that seems to be required for the build.

File: recipes/chemfiles-python/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  doc_url: http://chemfiles.github.io/chemfiles.py/latest/
+  dev_url: https://github.com/chemfiles/chemfiles.py
+
+build:
+  script:  # [not win]
+      - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
+      - python setup.py install # [not win]
+  script:  # [win]
+      - "%PYTHON%" setup.py install # [win]
+      - if errorlevel 1 exit 1  # [win]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: This line can be dropped.

File: recipes/chemfiles-python/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
+  summary: Python binding to chemfiles, a modern library for chemistry file reading and writing
+  doc_url: http://chemfiles.github.io/chemfiles.py/latest/
+  dev_url: https://github.com/chemfiles/chemfiles.py
+
+build:
+  script:  # [not win]
+      - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
+      - python setup.py install # [not win]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Please drop the selector and add add the arguments ` --single-version-externally-managed --record=record.txt`.

File: recipes/chemfiles-python/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
+  summary: Python binding to chemfiles, a modern library for chemistry file reading and writing
+  doc_url: http://chemfiles.github.io/chemfiles.py/latest/
+  dev_url: https://github.com/chemfiles/chemfiles.py
+
+build:
+  script:  # [not win]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Please drop this selector.

File: recipes/chemfiles-python/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+      - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Why is this needed? We often set the deployment target with the `toolchain`.

File: recipes/chemfiles-lib/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  doc_url: http://chemfiles.github.io/chemfiles/latest/
+  dev_url: https://github.com/chemfiles/chemfiles
+
+build:
+  script:
+    - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
+    - cmake -DCMAKE_INSTALL_PREFIX=$PREFIX -DENABLE_NETCDF=ON .
+    - cmake --build . --config release --target install
+    - cmake -DBUILD_SHARED_LIBS=ON .
+    - cmake --build . --config release --target install
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: For more complex scripts, we like to break them out into `build.sh` and `bld.bat`.

File: recipes/chemfiles-lib/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
+  summary: Modern library for chemistry file reading and writing
+  doc_url: http://chemfiles.github.io/chemfiles/latest/
+  dev_url: https://github.com/chemfiles/chemfiles
+
+build:
+  script:
+    - export MACOSX_DEPLOYMENT_TARGET=""  # [osx]
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: Why is this needed?

File: recipes/chemfiles-lib/meta.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{% set version = "0.6.0" %}
+{% set sha256 = "4e43dbe1b1d4c3a2eae923fecc3452d2b207efc1d511f6683164d51d22b0255f" %}
+
+package:
+  name: chemfiles-lib
+  version: {{ version }}
+
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
```
Comment (jakirkham) [inline recipes/chemfiles-lib/meta.yaml]: Please add `license_file: LICENCE.txt` to include the license file with the package.

File: recipes/chemfiles-python/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,48 @@
+{% set version = "0.6.0" %}
+{% set sha256 = "913a9ca911287e342918458619c29dfd094f4b25f7d6866048b9968fc63be207" %}
+
+package:
+  name: chemfiles-python
+  version: {{ version }}
+
+about:
+  home: http://chemfiles.github.io
+  license: MPL-v2.0
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Please add `license_file: LICENCE.txt` to include the license file with the package.

File: recipes/chemfiles-python/meta.yaml [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+requirements:
+  build:
+    - python
+    - enum34  # [py < 34]
+    - numpy
+    - chemfiles-lib ==0.6.0
+  run:
+    - python
+    - enum34  # [py < 34]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Please drop the spaces around `<`.

File: recipes/chemfiles-python/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+test:
+  commands:
+    - cd $SRC_DIR/tests
+    - python -m unittest discover
+
+requirements:
+  build:
+    - python
+    - enum34  # [py < 34]
```
Comment (jakirkham) [inline recipes/chemfiles-python/meta.yaml]: Please drop the spaces around `<`.

**Comment by jakirkham** [comment]
Thanks for proposing these, @Luthaf. Glad to see you over here. 😄

Tried to add some addition comments in addition to what the linter included. Once we get these cleaned up a bit more, we can take a closer look.

File: recipes/chemfiles-lib/meta.yaml (hunk 9 above)
Comment (Luthaf) [inline recipes/chemfiles-lib/meta.yaml]: The point of using cmake here is to have exactly the same commands to build the code in windows and *nix, and so to remove the need for separated build.sh and build.bat files.

File: recipes/chemfiles-lib/meta.yaml (hunk 10 above)
Comment (Luthaf) [inline recipes/chemfiles-lib/meta.yaml]: I do not remember, but I got an error on Os X when building the recipe a few months ago. I'll check again.

File: recipes/chemfiles-python/meta.yaml (hunk 4 above)
Comment (Luthaf) [inline recipes/chemfiles-python/meta.yaml]: Yes they are, because the setup.py `import chemfiles` to get the library version.

**Comment by Luthaf** [comment]
I addressed most of the comments; still have to enable tests for chemfiles-lib and check for what is going on with `export MACOSX_DEPLOYMENT_TARGET=""  # [osx]`.

**Comment by jakirkham** [comment]
I know the linter is mentioning tests. There are a couple types of tests we like to see.

Running the test suite (mainly for compiled code) so `make check`, `ctest` or similar. Normally that happens in the build.

For Python libraries import tests in the `test` section are nice.

Some checks to make sure files (libraries, headers, configuration files, etc.) were installed where they should be.

**Comment by Luthaf** [comment]
I actually need to `export MACOSX_DEPLOYMENT_TARGET=""`, because without it I get a Bus Error (segfault) in one of my tests. I will try to debug this using the binary built by conda, but I cannot reproduce with a fresh build.

**Comment by Luthaf** [comment]
The Circle-CI failure is strange: I got an error about `netcdf.h` file not found, even though `libnetcdf` is part of the `tests.require` entry. Any idea on the origin of this failure?

**Comment by Luthaf** [comment]
So after investigations (see the linked issue), the bus error is very weird. I am in favor of just unsetting `MACOSX_DEPLOYMENT_TARGET` in the build, since that make the bug disappear.

@jakirkham, I do not get the origin of the failure on travis/circle-ci, any idea? The Appveyor issue looks strange and related to installing libnetcdf with MSVC.

**Comment by Luthaf** [comment]
Appveyor failure seems legitimate, the build log is 

```
conda.resolve.Unsatisfiable: The following specifications were found to be in conflict:
  - cmake
  - libnetcdf 4.4.* -> feature:vc10
  - libnetcdf 4.4.* -> feature:vc14
  - libnetcdf 4.4.* -> feature:vc9
Use "conda info <package>" to see the dependencies for each package.
Command exited with code 1
```

I do not know what I can do here.

**Comment by Luthaf** [comment]
Ping @jakirkham? I finally disabled the Windows version, it can be added later.

**Comment by jakirkham** [comment]
So I'm in favor of merging.

Though I'd like to save us some time restarting CIs. Could we please put the non-meta recipes (looks like the first 2 commits in a separate PR)? Once done please ping me and I will merge ASAP.

Then we can cut this down the metapackage, which we can merge after the first two are built and deployed.

Sound good?

**Comment by Luthaf** [comment]
Sound good, I've separated the non-meta recipes in #1764

**Comment by jakirkham** [comment]
Thanks. Have gone ahead and merged that. It might be a little bit as conversion to feedstocks occurs on Travis CI ATM and they are a bit backed up as we speak.

**Comment by jakirkham** [comment]
So we ran into some issues with `chemfiles-python`. Have added PR ( https://github.com/conda-forge/chemfiles-python-feedstock/pull/1 ) to fix them. Please take a look. Once that is sorted we should be in good shape to proceed with this.

**Comment by jakirkham** [comment]
Toggling for CI.

**Comment by Luthaf** [comment]
Should I rebase and only include the meta recipe here?

**Comment by jakirkham** [comment]
That sounds like a good idea. Sure let's do that.

File: recipes/chemfiles/meta.yaml [hunk 15]
```diff
@@ -0,0 +1,28 @@
+{% set version = "0.6.0" %}
+
+package:
+  name: chemfiles
+  version: {{ version }}
+
+build:
+  number: 0
```
Comment (jakirkham) [inline recipes/chemfiles/meta.yaml]: Please add `skip: true  # [win]`. That should fix the CI failure. Also feel free to add `[skip appveyor]` in your commit message to avoid having to run that CI as it is not used.

**Comment by Luthaf** [comment]
Travis failure looks unrelated (502 gateway error from conda.anaconda.org), could you restart the build?

**Comment by jakirkham** [comment]
Yep. Restarted.

**Comment by jakirkham** [comment]
Thanks @Luthaf .
